Document user namespace changes in v1.27 (KEP-127) (#39860)
kubernetes/website: content/en/docs/concepts/workloads/pods/user-namespaces.md

Reviewer: It might be helpful to wait for the changes #40264 to land  in _dev-1.27_ and then rebase this PR.

However, until that's happened, **this** PR needs to:
```diff
-{{% thirdparty-content single="true" %}}
-<!-- if adding another runtime in the future, omit the single setting -->
+{{% thirdparty-content %}}
```

Author: Fixed, thanks

{{% thirdparty-content single="true" %}}
<!-- if adding another runtime in the future, omit the single setting -->

This is a Linux-only feature and support is needed in Linux for idmap mounts on
the filesystems used. This means:

> * The filesystem you use for `/var/lib/kubelet/pods/`, or the custom directory

Reviewer: ```suggestion
* On the node, the filesystem you use for `/var/lib/kubelet/pods/`, or the custom directory
```

Author: Fixed, thanks

  you configure for this, needs idmap mount support.

> * All the filesystems used in the pod need support for idmap mounts.

Reviewer: ```suggestion
* Every volume that is mounted into any container in the pod must support idmap mounts.
```

Author: This smells slightly incorrect. What supports idmap mount is the filesystem, not the volume. It doesn't matter if it is an EBS volume, a local disk volume or what. The only thing that matters is that the filesystem used in that volume supports idmap mounts in the kernel running on the host.

So, what about this instead?
```
* All the filesystems used in the pod's volumes must support for idmap mounts.
```

Author: Changed it to that, let me know what you think

{{% thirdparty-content single="true" %}}
<!-- if adding another runtime in the future, omit the single setting -->

This is a Linux-only feature and support is needed in Linux for idmap mounts.
This means:

* The filesystem you use for `/var/lib/kubelet/pods/`, or the custom directory

>   you configure for this, needs idmap mount support

Reviewer 2: ```suggestion
  you configure for this, needs idmap mount support.
```

Author: Fixed, thanks

* All the filesystems used in the pod need support for idmap mounts.

In practice this means you need at least Linux 6.3, as tmpfs started supporting
idmap mounts in that version. This is usually needed as several Kubernetes

> features use tmpfs (the service account token that is mounted by default uses a

Reviewer 2: ```suggestion
features use tmpfs (the ServiceAccount tokens that are mounted by default use a
```

Author: Hmm, are you sure? There is only one `token` file in the service account that is mounted:

```
# ls /run/secrets/kubernetes.io/serviceaccount/
ca.crt	namespace  token
```

And the token file seems to contain one token

Reviewer 2: I was trying to talk about the generals. In implementation, we are only mounting one ServiceAccount token today, even that token can be ignored.

> tmpfs, secrets use a tmpfs, etc.)

Reviewer 2: ```suggestion
tmpfs, Secrets use a tmpfs, etc.)
```

Some popular filesystems that support idmap mounts in Linux 6.3 are: btrfs,
ext4, xfs, fat, tmpfs, overlayfs.

In addition, support is needed in the 
{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}
to use this feature with Kubernetes stateless pods:

> * CRI-O: supports user namespaces with all Kubernetes versions >= v1.25.

Author: @giuseppe Can you confirm this is the case?

 * downwardAPI
 * emptyDir

## {{% heading "whatsnext" %}}

> * Take a look at the [user namespace example task](/docs/tasks/configure-pod-container/user-namespaces/)

Reviewer: ```suggestion
* Take a look at [Use a User Namespace With a Pod](/docs/tasks/configure-pod-container/user-namespaces/)
```

Author: Fixed, thanks

 * downwardAPI
 * emptyDir

To guarantee that the pod can read the files of such volumes, volumes are
created as if you specified `.spec.securityContext.fsGroup` as `0` for the Pod.
If it is specified to a different value, this other value will of course be
honored instead.
## {{% heading "whatsnext" %}}

As a by-product of this, folders and files for these volumes will have
permissions for the group, even if `defaultMode` or `mode` to specific items of
the volumes were specified without permissions to groups. For example, it is not
possible to mount these volumes in a way that its files have permissions only

> for the owner.

Reviewer 2: Why removing the above two paragraphs?

Author: See the commit description, that explains it:

```
Also, with the new implementation, the fsGroup was dropped. So removed
the mention to those limitations.
```
